dracut: add afterburn dracut module

yuqi-zhang opened the pull request:
Add dracut module to run afterburn in the initramfs, and add an
initial service to fetch hostname on necessary providers (Azure).

ashcrow reviewed (commented):
Looks sane to me

jlebon commented on dracut/30afterburn/afterburn-hostname.service:
```diff
@@ -0,0 +1,14 @@
+ConditionKernelCommandLine=|ignition.platform.id=azure
+ConditionFirstBoot=yes
+After=ignition-disks.service
+Before=ignition-files.service
```
Minor: maybe a comment there that we're purposely ordering before `ignition-files.service` so that the latter can still override the hostname if wanted?

yuqi-zhang replied:
Added some comments

jlebon reviewed (commented):
Just one small comment, otherwise LGTM!

lucab commented on dracut/30afterburn/afterburn-hostname.service:
```diff
@@ -0,0 +1,14 @@
+# Azure needs to fetch the hostname in the initramfs
+ConditionKernelCommandLine=|ignition.platform.id=azure
+ConditionFirstBoot=yes
+After=ignition-disks.service
```
@yuqi-zhang did you test this service unit on an actual Azure machine? Because I think I made a mistake and this should be `After=ignition-mount.service`.

jlebon replied:
I guess since it only writes to `/sysroot/etc`, technically it just needs `After=initrd-root-fs.target`. That'll work on both FCOS and RHCOS (which doesn't have `ignition-mount.service` yet).

yuqi-zhang replied:
I added `After=initrd-root-fs.target`
`Wants=initrd-root-fs.target` based on https://github.com/coreos/ignition-dracut/blob/spec2x/dracut/30ignition/coreos-static-network.service

lucab replied:
Ack. And do we care about `ignition-remount-sysroot.service` here? (I'm not so familiar with the new Ignition flow in initramfs)

jlebon replied:
> And do we care about `ignition-remount-sysroot.service` here?

It doesn't hurt, but shouldn't make a difference either, see: https://github.com/coreos/ignition-dracut/pull/38#issuecomment-451965981.

yuqi-zhang replied:
I suppose for safety we should? I'm also not sure on which systems that one actually runs.
And to your earlier point, I didn't test on Azure since I didn't have creds at the time, I made a mock unit with the same dependency tree to run on metal, and it worked as expected.

yuqi-zhang replied:
Added

jlebon commented on dracut/30afterburn/afterburn-hostname.service:
```diff
@@ -0,0 +1,24 @@
+# overwrite the configured hostname with e.g. a static one
+After=ignition-disks.service
+Before=ignition-files.service
+
```
This is maybe more personal preference, but I like to cut down the `After`/`Before` directives to what is actually necessary to make it clearer what the dependencies really are. So e.g. here, we should *only* need `After=initrd-root-fs.target` and `Before=ignition-files.service`. (We don't need `After=ignition-disks.service`, since clearly that must happen before `/sysroot` is mounted, and we don't need `Wants=initrd-root-fs.target` since that's already part of the transaction.)

That said, if you've already validated this works, I'm cool rolling with it like that as well!

yuqi-zhang replied:
I agree, I just quickly added those in. Let me clean it up a bit, thanks for the suggestion!

yuqi-zhang replied:
On that note, do you know where we're using https://github.com/coreos/ignition-dracut/blob/spec2x/dracut/30ignition/ignition-remount-sysroot.service ? Is that something we should look out for

jlebon replied:
We don't need it in RHCOS and FCOS, see https://github.com/coreos/ignition-dracut/commit/3ec0b39106a0724982da64dc348a38da7a471467. It was added back for other distros, see https://github.com/coreos/ignition-dracut/issues/37 and https://github.com/coreos/ignition-dracut/pull/38#issuecomment-451965981.

jlebon commented on dracut/30afterburn/afterburn-hostname.service:
```diff
@@ -0,0 +1,24 @@
+
+# This will be firstboot only to set up the correct
+# hostname for kube.
+ConditionFirstBoot=yes
```
One question, does this actually work correctly in the initrd? For the real root, systemd keys off of `/etc/machine-id`, but here it has to be able to make that determination before `/sysroot` is even mounted. So does it key off of the *initrd*'s `/etc/machine-id`? Except that check will never change result unless we regen the initrd.

An alternative here is to hook into `ignition-complete.target`, which was made for this: https://github.com/coreos/ignition-dracut/blob/df88988f2f0791ca9ea1e14298c8523501d980c5/dracut/30ignition/ignition-complete.target#L3.

For RHCOS, since it doesn't have that target yet, you could do e.g. `ignition-files.service.requires`?

ajeddeloh replied:
Afterburn does not depend on Ignition though, so we probably shouldn't make it use that target. I agree though that using `ConditionFirstBoot` from the initramfs seems sketchy. Not sure the best path forward here. I think this is a case where the distro-dependence is showing.

yuqi-zhang replied:
> Afterburn does not depend on Ignition though, so we probably shouldn't make it use that target.

We explicitly reference ignition targets, I don't really see a harm in cross-linking per se? Unless you want them to be as separate as possible.

I did some testing with a metal-bios image. If I `rd.break` into the initramfs on the first boot, I see the unit as inactive, and a message that: `Condition: start condition failed` (and same with subsequent boots). Interesting if I try to `systemctl start` the process manually, it will add a line `CondtionFirstBoot=yes was not met`. I'm not sure if that's the expected behaviour or not (i.e. which condition is not being met).

In either case, this does not seem to be very robust... should we

jlebon replied:
> Afterburn does not depend on Ignition though, so we probably shouldn't make it use that target.

I'd say this unit is inherently dependent on Ignition already by simply having `ConditionKernelCommandLine=|ignition.platform.id=azure`. :)

We could do something fancy here and "connect" the two in the overlay. Though unless there's demand for running Afterburn without Ignition, I'm not sure if it's worth the investment.

ajeddeloh replied:
I think this is fine now that I think about it. Part of me still just wants one giant fcos-dracut repo, but I think that ship has sailed.

jlebon commented on dracut/30afterburn/module-setup.sh:
```diff
@@ -0,0 +1,24 @@
+        "$systemdutildir/system/afterburn-hostname.service"
+
+    # We want the afterburn-hostname to be firstboot only, so ignition-provided
+    # hostname changes do not get overwritten on subsequentboots
```
Minor: `subsequent boots`
Super minor: `Ignition`

yuqi-zhang replied:
Fixed up, thanks for the review!

jlebon reviewed (commented):
Just some trivial nits, otherwise LGTM!

yuqi-zhang committed:
dracut: add afterburn dracut module

Add dracut module to run afterburn in the initramfs, and add an
initial service to fetch hostname on necessary providers (Azure).

Signed-off-by: Yu Qi Zhang <jerzhang@redhat.com>

jlebon reviewed (approved):
Yup this LGTM!
I don't have merge access, will leave it to someone else to have a final look and merge.
